File: 문제풀이/프로그래머스_고득점_KIT/해시/전화번호_목록.py

def solution(phone_book):
    tree = {}
    for phone_num in phone_book:
        cur = tree
        is_prefix = True
        for num in phone_num:
            if num not in cur:
                cur[num] = {}
                is_prefix = False
            else:
                if len(cur[num]) == 0:
                    return False
            cur = cur[num]
        if is_prefix:
            return False

    return True

'''
풀이 시간: 19분
접근:
    시간 복잡도 때문에 2중 for문 검사는 불가능하다.
    선형적으로 검사해야 함.
후기:
    생각보다 까다로웠다..

    또한 접두사가 될 수 있는 전화번호가 먼저 사전에 들어가는 경우를 체크 했어야 했다.

'''

File: CPP/프로그래머스_고득점_KIT/해시/전화번호_목록.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

using Node = map<int, void*>;

bool check_prefix(vector<string>& phone_book) {
    Node* tree = new Node;
    for (auto& phone_num : phone_book) {
        Node* cur = tree;
        bool is_prefix = true; // 현재 phone_num이 prefix인지
        for (auto& num : phone_num) {
            if (cur->find(num) == cur->end()) { // 원소가 없는 경우
                (*cur)[num] = new Node;
                is_prefix = false;
            }
            else {
                if (((Node *)(*cur)[num])->size() == 0) return true; // 현재 phone_num의 접두사를 찾은 경우
            }
            cur = (Node*)(*cur)[num];
        }
        if (is_prefix) return true;
    }
    return false;
}


bool solution(vector<string> phone_book) {
    bool answer = !check_prefix(phone_book);
    return answer;
}

/*
풀이 시간: 25분
접근:
    정렬을 하지 않고 map을 이용해 풀어보자.
    재귀적 구조를 만들 수 있을까?
후기:
    파이썬으로 이미 풀어본 문제임에도 불구하고 방법을 까먹어서 다시 생각하는데 조금 걸렸고..
    is_prefix 변수의 역할에 대해 중간에 혼동이 왔었다. -> 주석을 잘 달자.

    그리고 포인터 캐스팅 과정에서.. 괄호를 한 쌍 빼먹어서 오류가 났었다. 
    -> 뭔가 해결이 잘 안 되었었는데, 해당 부분의 코드를 지우고 다시 작성했더니 해결됐다.
*/